### DISCUSSION TITLE
Protocol 188 - Decoding Raw Data

### DISCUSSION THREAD

**Discussion by Fayrewood** [merbanan/rtl_433#3026] [Q&A]
I have successfully configured RTL_433 to pick up raw data from my Marlec Solar iBoost+ unit.

Is there an existing mechanism of decoding the raw data from protocol 188, so that I can create sensors in Home Assistant from the MQTT messages?

I am waiting for components to arrive to build an ESPHome device by JNSwanson, but have been playing with RTL_433 and got ahead of myself and thought it may be easier just to pass use decoded traffic from RTL_433.

```
[rtl_433] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[rtl_433] 
[rtl_433] time      : 2024-08-09T14:29:53.023021+0100
[rtl_433] model     : Marlec-Solar
[rtl_433] Raw data  : 2596cdabdbc6820ef800803f
[rtl_433] Integrity : CRC
[rtl_433] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[rtl_433] 
[rtl_433] time      : 2024-08-09T14:33:21.291884+0100
[rtl_433] model     : Marlec-Solar
[rtl_433] Raw data  : 25a9cdabd7e7820ef800803f
[rtl_433] Integrity : CRC
[rtl_433] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[rtl_433] 
[rtl_433] time      : 2024-08-09T14:42:14.309077+0100
[rtl_433] model     : Marlec-Solar
[rtl_433] Raw data  : 42eedcae74623107c51b0075b6633107c51b00040d3b04350c803fe946
[rtl_433] Integrity : CRC
```

**Comment by Fayrewood** [comment]
I have found https://github.com/merbanan/rtl_433/blob/master/src/devices/marlec_solar.c which explains nicely how to run the command line options required for the decoder.

I am running RTL_433 as a Home Assistant Add On and I have added 
```decoder n=Marlec,m=FSK_PCM,s=20,l=20,g=350,r=600,preamble=aad391d391``` in my config file.

The logs now show:

```
[rtl_433] 
[rtl_433] time      : 2024-08-09T15:02:14.595504+0100
[rtl_433] model     : Marlec
[rtl_433] count     : 1
[rtl_433] num_rows  : 1
[rtl_433] rows      : 
[rtl_433] 
[rtl_433] len       : 121
[rtl_433] data      : 0c255acdab0f45820e4a00803f693e0
[rtl_433] codes     : {121}0c255acdab0f45820e4a00803f693e0
[rtl_433] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[rtl_433] 
[rtl_433] time      : 2024-08-09T15:02:14.595504+0100
[rtl_433] model     : Marlec-Solar
[rtl_433] Raw data  : 255acdab0f45820e4a00803f
[rtl_433] Integrity : CRC
[rtl_433] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[rtl_433] 
[rtl_433] time      : 2024-08-09T15:02:14.599473+0100
[rtl_433] model     : Marlec
[rtl_433] count     : 1
[rtl_433] num_rows  : 1
[rtl_433] rows      : 
[rtl_433] 
[rtl_433] len       : 116
[rtl_433] data      : 0c2b6b36ac3d160838f80200ffe4a
[rtl_433] codes     : {116}0c2b6b36ac3d160838f80200ffe4a
[rtl_433] _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
[rtl_433] 
[rtl_433] time      : 2024-08-09T15:02:14.603440+0100
[rtl_433] model     : Marlec
[rtl_433] count     : 1
[rtl_433] num_rows  : 1
[rtl_433] rows      : 
[rtl_433] 
[rtl_433] len       : 117
[rtl_433] data      : 0c255acdab0f45820e3101007e6ae0
[rtl_433] codes     : {117}0c255acdab0f45820e3101007e6ae0
```

Feels like I'm making progress, but any idea how I get the data decoded?

**Reply by zuckschwerdt** [reply]
There is also #1739 with some info and you can run the decoding on raw data codes too, e.g. `rtl_433 -y aaaaaaaad391d3910c2555cdab855e820ef800803f7877` (note the prepended `aaaaaaaad391d391` and the included CRC at the end).

**Reply by Fayrewood** [reply]
Thanks!  which bits are the CRC and is that a constant?

**Reply by zuckschwerdt** [reply]
E.g. that example contains:
```
aaaaaaaa d391d391 0c 2555cdab855e820ef800803f 7877
Preamble Syncword Len      Payload            CRC
```
Only the payload will be in the "raw data" output, the CRC is a checksum and not constant.
